### ISSUE TITLE
Add indexes to models

### ISSUE THREAD

**Issue by PonteIneptique** [Capitains/Hook#39]
This morning, Hook did not reply anymore on some repo (namely greeklit)
This is because of indexes apparently.

Command to add index :

```shell
db.repository.createIndex({owner:1, name:1}) # Compound one for repository
db.repository.createIndex({authors:1})

db.repo_test.createIndex({run_at:-1})
db.repo_test.createIndex({uuid:1})
db.repo_test.createIndex({repository:1, branch:1})
```

**Comment by PonteIneptique** [comment]
Logs
```
[Wed Oct 28 07:28:26.472123 2015] [:error] [pid 8836:tid 140500921149184] [client 176.183.68.131:39294]     error_object), referer: http://ci.perseids.org/repo/PerseusDL/canonical-farsiLit
[Wed Oct 28 07:28:26.472143 2015] [:error] [pid 8836:tid 140500921149184] [client 176.183.68.131:39294] pymongo.errors.OperationFailure: database error: Executor error: Overflow sort stage buffered data usage of 35612436 bytes exceeds internal limit of 33554432 bytes, referer: http://ci.perseids.org/repo/PerseusDL/canonical-farsiLit
```

**Comment by PonteIneptique** [comment]
@fbaumgardt I'd like to have your feedback on the indexes for Mongo... You can find the model [here](https://github.com/Capitains/Hook/blob/master/Hook/models.py) as a start, and then we can discuss it ?

**Comment by fbaumgardt** [comment]
I’m not enough familiar with the exact processes in Hook Test to be commenting on the model itself, but here are 3 observations:

1. I don’t see any indices. Make sure that you have indices when you have larger collections and you’re doing searches on a field.
2. Possibly it makes sense to reference user objects from the user field in RepoTest.
3. Make sure you’re following best practices in handling the access tokens (I don’t know what those are, but I’d guess read-only on the data repos is enough rights?)

Feel free to call me on hangouts, I will be online in a bit.

**Comment by PonteIneptique** [comment]
Thanks @fbaumgardt 
Could you point me on a solid tutorial regarding indexes ?

**Comment by fbaumgardt** [comment]
Something MongoEngine-specific: http://docs.mongoengine.org/guide/defining-documents.html#indexes <http://docs.mongoengine.org/guide/defining-documents.html#indexes>

Something general: https://www.compose.io/articles/mongodb-indexing-best-practices/ <https://www.compose.io/articles/mongodb-indexing-best-practices/> & https://www.compose.io/articles/mongodb-and-indexing-best-practices-redux/ <https://www.compose.io/articles/mongodb-and-indexing-best-practices-redux/>

I’d look for three main concerns in deciding whether to index a field:

1. How often do you search on a field? (Many searches = large benefit from indexing)
2. How often do you write? (Many writes = DB always busy re-indexing)
3. Size of the collection? (large collection = long time for ad-hoc / un-indexed searches)

HTH!

**Comment by PonteIneptique** [comment]
Thanks again @fbaumgardt !

But HTH = Hello To Harry ? Hochimin To Houston ? Hanover's Tools Halo ?

**Comment by fbaumgardt** [comment]
Hope that helps! :P

**Comment by PonteIneptique** [comment]
![http://www.reactiongifs.com/r/Un5F2vt.gif](http://www.reactiongifs.com/r/Un5F2vt.gif)

**Comment by PonteIneptique** [comment]
Fixed as of @3ba2a5e10419024f5df788b1950b56fe1607b096
Created a technical debt doing so.

**Comment by fbaumgardt** [comment]
@ponteineptique FYI, the MongoDB course that I sent you a while ago has a lecture on indexes this week, which might be of interest to you. Here’s the first in a series of videos: https://www.youtube.com/watch?v=U3iWPF5jP-g <https://www.youtube.com/watch?v=U3iWPF5jP-g>
